File: cpp/Mod_absolute.cpp
//int absolute(int a) {return a < 0 ? -a : a;}

#pragma once


Int absolute(Int a);
Int absolute_debug(Int a);

class absolute_class {
    
    //VARIABLES        
    Int a;
    Int a_spnew;
    Int res;
    Int res_spnew;
    bool debug_spv;
    
    //PREDICATES


    //OPERATIONS
    void o_init() {                    
        res_spnew = sp_if(sp_lt(a, 0), sp_minus(0, a), a);
        
        res = res_spnew;       

        if(debug_spv) {
            cout << "o_init" << " ; "  
             << "" << ";"  << res  
            << endl;   
        }

        return;
    }
    
    public:
    absolute_class(Int constr_a) : a(constr_a)   {  }
        
    Int execute(bool debug) {           
        debug_spv = debug;
        
        if(debug_spv) {
            cout << "absolute" << ";"  
            << "a"<< "; " << "res" 
            << endl;             
            
            cout << "" << "; "  
             << a << ";"  << ""  
            << endl;   
        }
        
        o_init();           
        return res;
    }
        
    
};

Int absolute(Int a) {
    absolute_class obj(a);  
    return obj.execute(false);
}


Int absolute_debug(Int a) {
    absolute_class obj(a);  
    return obj.execute(true);
}

File: hs/Mod_absolute.hs
{-module Mod_absolute ( absolute ) where
import Prelude hiding (and, or)
import Debug.Trace
import Splib
import Types

absolute :: Int -> Int
absolute x  | x < 0 = -x
            | x >= 0 = x-}


module Mod_absolute ( absolute ) where
import Prelude hiding (and, or)
import Debug.Trace
import Splib
import Types


trace' :: String -> a -> a
--trace' str x = (trace str x)
trace' str x = x

data Data_absolute  = Data_absolute  {
  a :: Int,
  res :: Int
}

--SIGNATURES
absolute :: Int -> Int
o_init :: Data_absolute -> Int


--DEFINITIONS
absolute a = (o_init (Data_absolute a undefined))



o_init (Data_absolute a res) = (trace' "o_init" flow)
  where
    res' = (sp_if (sp_lt a 0) (sp_minus 0 a) a)
    a' = a
    data' = (Data_absolute a' res')
    flow =
      res'
